File: go/nucleus/retry_test.go
package nucleus

import (
	"errors"
	"fmt"
	"testing"

	"github.com/jackc/pgx/v5/pgconn"
)

func pgErr(code string) error {
	return &pgconn.PgError{Code: code, Message: "synthetic"}
}

// The whole point of these helpers is that they classify by SQLSTATE and not by
// message text. A driver surfaces the code; the message is free-form and
// changes. Nucleus itself shipped this bug twice — a 2PL kill reported as
// XX000, then its follow-up error reported as XX000 — so the client half of the
// contract gets a test.
func TestIsSerializationFailure(t *testing.T) {
	cases := []struct {
		name string
		err  error
		want bool
	}{
		{"40001 serialization failure", pgErr(SQLStateSerializationFailure), true},
		{"25P02 in failed transaction", pgErr(SQLStateInFailedTransaction), true},
		{"55P03 lock timeout is NOT retryable", pgErr(SQLStateLockNotAvailable), false},
		{"23505 unique violation", pgErr("23505"), false},
		{"XX000 internal", pgErr("XX000"), false},
		{"plain error", errors.New("boom"), false},
		{"nil", nil, false},
		{"wrapped 40001", fmt.Errorf("outer: %w", pgErr(SQLStateSerializationFailure)), true},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			if got := IsSerializationFailure(tc.err); got != tc.want {
				t.Fatalf("IsSerializationFailure(%v) = %v, want %v", tc.err, got, tc.want)
			}
		})
	}
}

// A lock timeout means the holder is still there. Retrying spins against a lock
// that is not moving, which turns one stuck transaction into a busy loop — so
// it must be distinguishable from a conflict, not lumped in with it.
func TestIsLockNotAvailable(t *testing.T) {
	if !IsLockNotAvailable(pgErr(SQLStateLockNotAvailable)) {
		t.Fatal("55P03 should be recognised as a lock timeout")
	}
	if IsLockNotAvailable(pgErr(SQLStateSerializationFailure)) {
		t.Fatal("40001 must not be classified as a lock timeout")
	}
	if IsSerializationFailure(pgErr(SQLStateLockNotAvailable)) {
		t.Fatal("a lock timeout must never be retried as a serialization failure")
	}
	if IsLockNotAvailable(errors.New("boom")) {
		t.Fatal("a non-pg error is not a lock timeout")
	}
}

// Wrapping matters because every layer between the driver and the application
// adds context with %w. Classification that only works on the bare driver error
// works nowhere real.
func TestClassificationSeesThroughWrapping(t *testing.T) {
	deep := fmt.Errorf("repo: %w", fmt.Errorf("service: %w", pgErr(SQLStateSerializationFailure)))
	if !IsSerializationFailure(deep) {
		t.Fatal("classification must unwrap")
	}
}

func TestDefaultRetryOptionsAreSane(t *testing.T) {
	o := DefaultRetryOptions()
	if o.MaxAttempts < 2 {
		t.Fatalf("MaxAttempts = %d; a retry helper that retries zero times is not one", o.MaxAttempts)
	}
	if o.BaseDelay <= 0 || o.MaxDelay < o.BaseDelay {
		t.Fatalf("nonsensical backoff: base=%v max=%v", o.BaseDelay, o.MaxDelay)
	}
}

File: typescript/packages/neutron-nucleus/src/retry.test.ts
import assert from 'node:assert/strict';
import { describe, it } from 'node:test';

import {
  IN_FAILED_TRANSACTION,
  LOCK_NOT_AVAILABLE,
  RetryExhaustedError,
  SERIALIZATION_FAILURE,
  isLockNotAvailable,
  isSerializationFailure,
  sqlState,
  withRetry,
} from './retry.js';

/** Minimal stand-in for a node-postgres error, which puts SQLSTATE on `.code`. */
class PgErr extends Error {
  code: string;
  constructor(code: string) {
    super(code);
    this.code = code;
  }
}

const noSleep = { baseDelayMs: 1, sleep: async () => {}, random: () => 0 };

describe('sqlState', () => {
  it('reads the SQLSTATE off a driver error', () => {
    assert.equal(sqlState(new PgErr('40001')), '40001');
  });

  it('walks the cause chain', () => {
    // Every layer between driver and application adds context with `cause`;
    // classification that only sees the bare error works nowhere real.
    const deep = new Error('repo', { cause: new Error('service', { cause: new PgErr('40001') }) });
    assert.equal(sqlState(deep), '40001');
  });

  it("ignores Nucleus's own symbolic codes", () => {
    // NucleusError.code is 'QUERY_ERROR' etc. Treating that as a SQLSTATE
    // would misclassify every wrapped error.
    assert.equal(sqlState({ code: 'QUERY_ERROR' }), undefined);
    assert.equal(sqlState({ code: 'CONNECTION_ERROR' }), undefined);
  });

  it('returns undefined when there is no code at all', () => {
    assert.equal(sqlState(new Error('boom')), undefined);
    assert.equal(sqlState(null), undefined);
  });
});

describe('classification', () => {
  const cases: Array<[string, boolean]> = [
    [SERIALIZATION_FAILURE, true],
    [IN_FAILED_TRANSACTION, true],
    // The one that must never be lumped in with conflicts.
    [LOCK_NOT_AVAILABLE, false],
    ['23505', false],
    // The code Nucleus itself wrongly used for a kill, twice.
    ['XX000', false],
  ];
  for (const [code, retryable] of cases) {
    it(`${code} retryable=${retryable}`, () => {
      assert.equal(isSerializationFailure(new PgErr(code)), retryable);
    });
  }

  it('separates a lock timeout from a conflict', () => {
    assert.equal(isLockNotAvailable(new PgErr(LOCK_NOT_AVAILABLE)), true);
    assert.equal(isLockNotAvailable(new PgErr(SERIALIZATION_FAILURE)), false);
  });
});

describe('withRetry', () => {
  it('retries a conflict and then succeeds', async () => {
    let calls = 0;
    const result = await withRetry(async () => {
      calls += 1;
      if (calls <= 2) throw new PgErr(SERIALIZATION_FAILURE);
      return 'done';
    }, noSleep);
    assert.equal(result, 'done');
    assert.equal(calls, 3);
  });

  it('gives up after maxAttempts and keeps the cause', async () => {
    let calls = 0;
    await assert.rejects(
      withRetry(async () => {
        calls += 1;
        throw new PgErr(SERIALIZATION_FAILURE);
      }, { ...noSleep, maxAttempts: 3 }),
      (err: unknown) => err instanceof RetryExhaustedError,
    );
    assert.equal(calls, 3);
  });

  it('attempts a lock timeout exactly once', async () => {
    // 55P03 means the holder is still there. Retrying turns one stuck
    // transaction into a busy loop against a lock that will not move.
    let calls = 0;
    await assert.rejects(
      withRetry(async () => {
        calls += 1;
        throw new PgErr(LOCK_NOT_AVAILABLE);
      }, { ...noSleep, maxAttempts: 5 }),
      (err: unknown) => err instanceof PgErr,
    );
    assert.equal(calls, 1);
  });

  it('propagates a non-retryable error unchanged', async () => {
    await assert.rejects(
      withRetry(async () => {
        throw new PgErr('23505');
      }, noSleep),
      (err: unknown) => (err as PgErr).code === '23505',
    );
  });

  it('does not retry a success', async () => {
    let calls = 0;
    await withRetry(async () => {
      calls += 1;
      return 1;
    }, noSleep);
    assert.equal(calls, 1);
  });
});
